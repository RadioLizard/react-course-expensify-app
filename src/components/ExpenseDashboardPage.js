import React from 'react'
import ConnectedExpenseList from './ExpenseList'
import ConnectedExpenseListFilters from './ExpenseListFilters'
import ConnectedExpenseSummary from './ExpenseSummary'

const ExpenseDashboardPage = () =>{
    return(
    <div>
        <ConnectedExpenseSummary/> 
        <ConnectedExpenseListFilters/>
        <ConnectedExpenseList/>
    </div>
    )
}

export default ExpenseDashboardPage