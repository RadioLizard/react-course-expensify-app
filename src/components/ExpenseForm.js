import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

class ExpenseForm extends React.Component{
    constructor(props){
    super(props)
    this.state={
        description: props.expense ? props.expense.description : '',
        amount: props.expense ? (props.expense.amount/ 100).toString() : 0,
        note: props.expense ? props.expense.note : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
        focused:false,
        error:undefined
    }
}
    onDescriptionChange=(e)=>{
        const description = e.target.value
        this.setState(()=>({description}))
    }
    onNoteChange=(e)=>{
        const note=e.target.value
        this.setState(()=>({note}))
    }
    onAmountChange=(e)=>{
        const amount=e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }
    onDateChange=(createdAt)=>{
        if(createdAt){
            this.setState(()=>({createdAt}))
        }
    }
    onFocusChange=({focused})=>{
        this.setState(()=>({focused}))
    }
    onSubmit=(e)=>{
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error: "Please add a valid description and amount."}))
        }
        else{
            this.setState(()=>({error:undefined}))
            this.props.onSubmit({
              description: this.state.description,
              amount: parseFloat(this.state.amount, 10)*100,
              createdAt: this.state.createdAt.valueOf(),
              note: this.state.note
            })
        }
    }
    render(){
        return(
            <div>
                {!!this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="number"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.focused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=>false}
                    />
                    <textarea
                        value={this.state.note}
                        placeholder="add a note for this expense"
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm