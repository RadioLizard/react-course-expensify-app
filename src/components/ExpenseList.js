import React from 'react'
import {connect} from 'react-redux'
import Expense from './Expense'
import getVisibleExpenses from '../selectors/expenses'

const ExpenseList=(props)=>(
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
        {
            props.expenses.length===0 ? (
                <div className="list-item list-item--message">
                    <p>No expenses to show</p>
                </div>
            ):(
                props.expenses.map((expense, index)=>(
                <Expense
                    {...expense}
                    key={index}
                />))
            )
        }
        </div>
    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenses: getVisibleExpenses(state.expenses, state.filters),
    }
}
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList
export {ExpenseList}