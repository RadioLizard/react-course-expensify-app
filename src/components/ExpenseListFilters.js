import React from 'react'
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters'

class ExpenseListFilters extends React.Component{
    state={
        calandarFocused: null,
    }
    onDatesChange=({startDate, endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange=(calandarFocused)=>{
        this.setState(()=>({calandarFocused}))
    }
    onTextChange=(e)=>{
        this.props.setTextFilter(e.target.value)
    }
    onSortChange=(e)=>{
        if(e.target.value==="date"){
            this.props.sortByDate()
        }
        else if(e.target.value==="amount"){
            this.props.sortByAmount()
        }
    }
    render(){
        return(
        <div> 
            <input 
                type="text" 
                value={this.props.filters.text} 
                onChange={this.onTextChange}
            />
            <p>Sort By:</p>
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}> 
                <option value="date">date</option>
                <option value="amount">amount</option>
            </select>
            <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calandarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
            />
        </div>
        )
    }
}

const mapStateToProps = (state) =>{
    return{
        filters:state.filters
    }
}

const mapDispatchToProps = (dispatch)=>({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

const ConnectedExpenseListFilters=connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)

export default ConnectedExpenseListFilters
export {ExpenseListFilters}