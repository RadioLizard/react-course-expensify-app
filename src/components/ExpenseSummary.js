import React from 'react'
import {connect} from 'react-redux'
import  numeral from 'numeral'
import selectExpensesTotal from '../selectors/filters'
import getVisibleExpenses from '../selectors/expenses'

const ExpenseSummary = (props) =>(
    <div>
        <p>Viewing {props.expenseCount} 
            {props.expenseCount===1 ? 
            <span> expense</span> : 
            <span> expenses</span>}, 
            which {props.expenseCount===1 ?
            <span>totals </span>:
            <span>total </span>}
            {numeral(props.expenseTotal / 100).format('$0,0.00')}</p>
    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenseTotal: selectExpensesTotal(getVisibleExpenses(state.expenses, state.filters)),
        expenseCount: getVisibleExpenses(state.expenses, state.filters).length
    }
}
const ConnectedExpenseSummary = connect(mapStateToProps)(ExpenseSummary)

export default ConnectedExpenseSummary
export{ExpenseSummary}