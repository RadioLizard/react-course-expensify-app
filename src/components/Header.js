import React from 'react'
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'
import {startLogout} from '../actions/auth'

const Header =({startLogout})=>(
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard"> 
                    <h1>Expensify</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
)   

const mapDispatchToProps=(dispatch)=>{
    return{
        startLogout: ()=> dispatch(startLogout())
    }
}
const ConnectedHeader =  connect(undefined, mapDispatchToProps)(Header)

export {Header, ConnectedHeader}