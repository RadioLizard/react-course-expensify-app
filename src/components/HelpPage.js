import React from 'react'

const HelpPage = () =>{
    return(
    <div>
        <p>Get some fucking help</p>   
    </div>
    )
}

export default HelpPage