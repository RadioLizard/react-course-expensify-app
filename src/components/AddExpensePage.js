import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense} from '../actions/expenses'

class AddExpensePage extends React.Component {
    addExpense=(expense)=>{
        this.props.startAddExpense(expense)
        this.props.history.push('/') 
    }
    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add an Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        onSubmit={this.addExpense}
                    />
                </div>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch)=>({
    startAddExpense: (expense)=> dispatch(startAddExpense(expense))
})
    
const ConnectedAddExpensePage=connect(undefined, mapDispatchToProps)(AddExpensePage)

export default ConnectedAddExpensePage
export{AddExpensePage}