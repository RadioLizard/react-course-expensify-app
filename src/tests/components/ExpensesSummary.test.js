import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseSummary} from '../../components/ExpenseSummary'

test('Should render Expense Summary',()=>{
    const wrapper=shallow(<ExpenseSummary expenseTotal={0} expenseCount={0}/>)
    expect(wrapper).toMatchSnapshot()
})

test('Should render Expense Summary correctly with just one item',()=>{
    const wrapper=shallow(<ExpenseSummary expenseTotal={134} expenseCount={1}/>)
    expect(wrapper).toMatchSnapshot()
})

test('Should render Expense Summary correctly with multiple items',()=>{
    const wrapper=shallow(<ExpenseSummary expenseTotal={134567} expenseCount={7}/>)
    expect(wrapper).toMatchSnapshot()
})