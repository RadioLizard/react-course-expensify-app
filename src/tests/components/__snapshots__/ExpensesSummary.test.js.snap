// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expense Summary 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing
      <span
        className="span"
      >
         
        0
         
      </span>
      <span>
         expenses
      </span>
      , which 
      <span>
        total 
      </span>
      <span
        className="span"
      >
        $0.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render Expense Summary correctly with just one item 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing
      <span
        className="span"
      >
         
        1
         
      </span>
      <span>
         expense
      </span>
      , which 
      <span>
        totals 
      </span>
      <span
        className="span"
      >
        $1.34
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should render Expense Summary correctly with multiple items 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing
      <span
        className="span"
      >
         
        7
         
      </span>
      <span>
         expenses
      </span>
      , which 
      <span>
        total 
      </span>
      <span
        className="span"
      >
        $1,345.67
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
